### ISSUE TITLE
App moves to the background state if you click it in the start menu while it's already running

### ISSUE THREAD

**Issue by rigdern** [microsoft/react-native-windows#1785]
Sometimes the `AppStateModule` thinks it's in the background even though the app is in the foreground.

### Repro Steps

1. Open the app from the start menu. Keep the app in the foreground.
2. Click on the app in the start menu again.

**Unexpected result:** The `AppStateModule` reports to JavaScript that the app is in the background even though it's in the foreground.

### Cause

Repro step (2) causes [`ReactApplication.OnLaunched`](https://github.com/Microsoft/react-native-windows/blob/179f6f6dac04cc7b4a31538445d43340da557903/ReactWindows/ReactNative/ReactApplication.cs#L57) to run. This method calls `OnCreate` which calls `_host.OnResume` which transitions RNW to the background state.

I didn't repro this with `ReactApplication` but I reproed this in my own app which uses very similar code from an older `App.xaml.cs` template

**Comment by rozele** [comment]
@rigdern can you take a look at #1799 and see if that resolves this issue?

**Comment by rigdern** [comment]
It looks like that would resolve the immediate issue in the template that I reported here.

I want to give you some insight on the challenges faced by real RNW apps in this area. We have more app entry points that need to call `OnResume` and we had several bugs where `OnResume` was over called (exactly like the issue I described here) before we realized a good pattern for using `OnResume`. Here are additional entry point scenarios we have:
  - `OnBackgroundActivated`: Called for in-process background tasks such as handling push notifications or a timezone change.
  - `OnShareTargetActivated`: Called when the user wants to share some content to your app when they're using another app. For this scenario, the OS creates a secondary window in which you render your UI.
  - [MyPeople](https://docs.microsoft.com/en-us/windows/uwp/contacts-and-calendar/my-people-support): Comes thru `OnActivated`. Enables you to draw your app's UI in the system address book. For this scenario, the OS creates a secondary window in which you render your UI.

A challenge with all of these is sometimes these act as app entry points (i.e. your process is launched in order to handle a background task, a share request, etc.) and other times they are invoked while the app is already running (i.e. the user is already using the app and then a background task or share request comes in). In the former case, `OnResume` needs to be called. In the latter case, `OnResume` should **not** be called.

As I mentioned, it took us a while before we figured out how to use `OnResume` properly. Bad things happen if you call it too little or too much. I think the lack of examples and the misleading name of `OnResume` contributed to the challenges. `OnResume` sounds like it should be called only in the app's `Resuming` lifecycle event. I know the template calls it in `OnActivated/OnLaunched` (before `OnCreate`! that's counterintuitive) which I found confusing. Eventually, we found it's more useful to think of the host's `OnResume` and `OnSuspend` methods as being for starting and stopping the host.

Instead of calling `OnResume` and `OnSuspend` directly, we settled on some helpers which assist us in starting and stopping the host correctly:

```c#
// Helpers that make it easier to use ReactNativeHost's `OnResume` and
// `OnSuspend` methods by improving upon their names and supporting
// redundant calls for `OnResume`.

// Must only be used from the main dispatcher.
private bool _isReactNativeHostStopped = true;

// Starts `ReactNativeHost` if it's not already running. Should be called
// from each app entry point and when the app is resumed from suspension.
// Must be called on the main dispatcher.
private void EnsureReactNativeHostStarted()
{
    if (_isReactNativeHostStopped)
    {
        RnLog.Info(_appTag, "EnsureReactNativeHostStarted: starting host");
        _host.OnResume(Exit);
        _isReactNativeHostStopped = false;
    }
    else
    {
        RnLog.Info(_appTag, "EnsureReactNativeHostStarted: host already started");
    }
}

// Should be called when the app is suspended.
// Must be called on the main dispatcher.
private void StopReactNativeHost()
{
    RnLog.Info(_appTag, "StopReactNativeHost");
    _host.OnSuspend();
    _isReactNativeHostStopped = true;
}
```

With this helper, we can unconditionally call `EnsureReactNativeHostStarted()` in app entry points like `OnBackgroundActivated` and be confident that it'll work properly regardless of whether the app was just launched to handle this request or if the app has already been running. Also, I think the new naming scheme makes the code easier to understand.
